File: generated_code/GPT4/Python/056_SplayTree.py
class SplayTree:
    class Node:
        __slots__ = ['key', 'left', 'right', 'parent']
        def __init__(self, key, parent=None):
            self.key = key
            self.left = None
            self.right = None
            self.parent = parent

    def __init__(self):
        self.root = None

    # ---------- Rotation helpers ----------
    def _rotate_left(self, x):
        y = x.right
        if y is None:
            return
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        else:
            if x == x.parent.left:
                x.parent.left = y
            else:
                x.parent.right = y
        y.left = x
        x.parent = y

    def _rotate_right(self, x):
        y = x.left
        if y is None:
            return
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        else:
            if x == x.parent.left:
                x.parent.left = y
            else:
                x.parent.right = y
        y.right = x
        x.parent = y

    # ---------- Splay operation ----------
    def _splay(self, x):
        while x.parent:
            p = x.parent
            gp = p.parent
            if not gp:  # Zig
                if x == p.left:
                    self._rotate_right(p)
                else:
                    self._rotate_left(p)
            else:
                if x == p.left and p == gp.left:
                    self._rotate_right(gp)
                    self._rotate_right(p)
                elif x == p.right and p == gp.right:
                    self._rotate_left(gp)
                    self._rotate_left(p)
                elif x == p.right and p == gp.left:
                    self._rotate_left(p)
                    self._rotate_right(gp)
                elif x == p.left and p == gp.right:
                    self._rotate_right(p)
                    self._rotate_left(gp)

    # ---------- Search with splay ----------
    def search(self, key):
        node = self.root
        last = None
        while node:
            last = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        if last:
            self._splay(last)
        return False

    # ---------- Insert ----------
    def insert(self, key):
        if self.root is None:
            self.root = self.Node(key)
            return

        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                self._splay(node)
                return  # Key already exists, do nothing
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        new_node = self.Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        self._splay(new_node)

    # ---------- Delete ----------
    def delete(self, key):
        if not self.search(key):
            return  # Key not found, nothing to delete

        node = self.root  # After search, node with key is at root
        # Remove root
        if node.left is None:
            self._replace_root(node.right)
        elif node.right is None:
            self._replace_root(node.left)
        else:
            # Find max in left subtree
            left_subtree = node.left
            left_subtree.parent = None
            max_left = left_subtree
            while max_left.right:
                max_left = max_left.right
            self._splay_node_in_subtree(max_left, left_subtree)
            # Now max_left is new root of left subtree
            max_left.right = node.right
            if node.right:
                node.right.parent = max_left
            self.root = max_left
            max_left.parent = None

    def _replace_root(self, node):
        self.root = node
        if node:
            node.parent = None

    def _splay_node_in_subtree(self, node, subtree_root):
        # Splay node to root of subtree (used in delete)
        while node.parent:
            if node.parent == subtree_root:
                if node == node.parent.left:
                    self._rotate_right(node.parent)
                else:
                    self._rotate_left(node.parent)
            else:
                p = node.parent
                gp = p.parent
                if node == p.left and p == gp.left:
                    self._rotate_right(gp)
                    self._rotate_right(p)
                elif node == p.right and p == gp.right:
                    self._rotate_left(gp)
                    self._rotate_left(p)
                elif node == p.right and p == gp.left:
                    self._rotate_left(p)
                    self._rotate_right(gp)
                elif node == p.left and p == gp.right:
                    self._rotate_right(p)
                    self._rotate_left(gp)

    # ---------- Utility ----------
    def __contains__(self, key):
        return self.search(key)

    def __iter__(self):
        stack = []
        node = self.root
        while stack or node:
            while node:
                stack.append(node)
                node = node.left
            node = stack.pop()
            yield node.key
            node = node.right

    def __len__(self):
        def count(node):
            if not node:
                return 0
            return 1 + count(node.left) + count(node.right)
        return count(self.root)

File: generated_code/GPT4/Java/056_SplayTree.java
public class SplayTree {
    private class Node {
        int key;
        Node left, right, parent;

        Node(int key) {
            this.key = key;
        }
    }

    private Node root;

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, does nothing.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }
        Node curr = root, parent = null;
        while (curr != null) {
            parent = curr;
            if (key < curr.key)
                curr = curr.left;
            else if (key > curr.key)
                curr = curr.right;
            else {
                splay(curr);
                return; // Key already in tree
            }
        }
        Node newNode = new Node(key);
        newNode.parent = parent;
        if (key < parent.key)
            parent.left = newNode;
        else
            parent.right = newNode;
        splay(newNode);
    }

    /**
     * Deletes a key from the splay tree if it exists.
     */
    public void delete(int key) {
        Node node = findNode(key);
        if (node == null)
            return;
        splay(node);
        if (node.left != null) {
            Node leftSubtree = node.left;
            leftSubtree.parent = null;
            Node maxLeft = subtreeMaximum(leftSubtree);
            splay(maxLeft);
            maxLeft.right = node.right;
            if (node.right != null)
                node.right.parent = maxLeft;
            root = maxLeft;
            root.parent = null;
        } else if (node.right != null) {
            root = node.right;
            root.parent = null;
        } else {
            root = null;
        }
    }

    /**
     * Searches for a key in the splay tree.
     * If found, splays the node to the root and returns the key.
     * If not found, returns null.
     */
    public Integer search(int key) {
        Node node = findNode(key);
        if (node != null) {
            splay(node);
            return node.key;
        }
        return null;
    }

    // Helper: Find node with given key, no splaying
    private Node findNode(int key) {
        Node curr = root;
        while (curr != null) {
            if (key < curr.key)
                curr = curr.left;
            else if (key > curr.key)
                curr = curr.right;
            else
                return curr;
        }
        return null;
    }

    // Helper: Splay the given node to the root
    private void splay(Node x) {
        if (x == null)
            return;
        while (x.parent != null) {
            Node p = x.parent;
            Node g = p.parent;
            if (g == null) {
                // Zig
                if (x == p.left)
                    rotateRight(p);
                else
                    rotateLeft(p);
            } else if (x == p.left && p == g.left) {
                // Zig-Zig
                rotateRight(g);
                rotateRight(p);
            } else if (x == p.right && p == g.right) {
                // Zig-Zig
                rotateLeft(g);
                rotateLeft(p);
            } else if (x == p.right && p == g.left) {
                // Zig-Zag
                rotateLeft(p);
                rotateRight(g);
            } else if (x == p.left && p == g.right) {
                // Zig-Zag
                rotateRight(p);
                rotateLeft(g);
            }
        }
        root = x;
    }

    // Helper: Rotate node x to left (x becomes left child of its right child)
    private void rotateLeft(Node x) {
        Node y = x.right;
        if (y == null)
            return;
        x.right = y.left;
        if (y.left != null)
            y.left.parent = x;
        y.parent = x.parent;
        if (x.parent == null)
            root = y;
        else if (x == x.parent.left)
            x.parent.left = y;
        else
            x.parent.right = y;
        y.left = x;
        x.parent = y;
    }

    // Helper: Rotate node x to right (x becomes right child of its left child)
    private void rotateRight(Node x) {
        Node y = x.left;
        if (y == null)
            return;
        x.left = y.right;
        if (y.right != null)
            y.right.parent = x;
        y.parent = x.parent;
        if (x.parent == null)
            root = y;
        else if (x == x.parent.left)
            x.parent.left = y;
        else
            x.parent.right = y;
        y.right = x;
        x.parent = y;
    }

    // Helper: Return the max node in the subtree rooted at node
    private Node subtreeMaximum(Node node) {
        while (node.right != null)
            node = node.right;
        return node;
    }

    // Optional: For debugging, in-order traversal
    /*
    public void printInOrder() {
        printInOrder(root);
        System.out.println();
    }

    private void printInOrder(Node node) {
        if (node != null) {
            printInOrder(node.left);
            System.out.print(node.key + " ");
            printInOrder(node.right);
        }
    }
    */
}